analyzers/tests/SonarAnalyzer.Test/Helpers/DisjointSetsPrimitivesTest.cs

```diff
@@ -0,0 +1,70 @@
+        FindRoot(parents, "1").Should().Be("1");                    // Reflexivity
+        Union(parents, "1", "2");
+        FindRoot(parents, "1").Should().Be(FindRoot(parents, "2")); // Correctness
+        Union(parents, "1", "2");
+        FindRoot(parents, "1").Should().Be(FindRoot(parents, "2")); // Idempotency
+        Union(parents, "2", "1");
+        FindRoot(parents, "1").Should().Be(FindRoot(parents, "2")); // Symmetry
+
+        FindRoot(parents, "3").Should().Be("3");
+        Union(parents, "2", "3");
+        FindRoot(parents, "2").Should().Be(FindRoot(parents, "3"));
+        FindRoot(parents, "1").Should().Be(FindRoot(parents, "3")); // Transitivity
+        Union(parents, "3", "4");
+        FindRoot(parents, "1").Should().Be(FindRoot(parents, "4")); // Double transitivity
+        Union(parents, "4", "1");
+        FindRoot(parents, "4").Should().Be(FindRoot(parents, "1")); // Idempotency after transitivity
+    }
+
+    [TestMethod]
+    public void DisjointSetsAndUnion_AreConsistent()
+    {
+        var parents = Enumerable.Range(1, 5).ToDictionary(i => i.ToString(), i => i.ToString());
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1"], ["2"], ["3"], ["4"], ["5"]]); // Initial state
+        Union(parents, "1", "2");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2"], ["3"], ["4"], ["5"]]);   // Correctness
+        Union(parents, "1", "2");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2"], ["3"], ["4"], ["5"]]);   // Idempotency
+
+        Union(parents, "2", "3");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3"], ["4"], ["5"]]);
+        Union(parents, "1", "3");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3"], ["4"], ["5"]]);     // Idempotency after transitivity
+
+        Union(parents, "4", "5");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3"], ["4", "5"]]);       // Separated trees
+        Union(parents, "3", "4");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3", "4", "5"]]);         // Merged trees
+    }
+
+    [TestMethod]
+    public void DisjointSetsAndUnion_OfNestedTrees()
+    {
+        var parents = Enumerable.Range(1, 6).ToDictionary(i => i.ToString(), i => i.ToString());
+        Union(parents, "1", "2");
+        Union(parents, "3", "4");
+        Union(parents, "5", "6");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2"], ["3", "4"], ["5", "6"]]); // Merge of 1-height trees
+        Union(parents, "2", "3");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3", "4"], ["5", "6"]]);   // Merge of 2-height trees
+        Union(parents, "4", "5");
+        DisjointSets(parents).Should().BeEquivalentTo<List<string>>([["1", "2", "3", "4", "5", "6"]]);     // Merge of 1-height tree and 2-height tree
```

Comment:
To remove clutter, you could extract:

```csharp
private static void DisjointSetShouldBe(Dictionary<string, string> parents, List<List<string>> expected) =>
  DisjointSets(parents).Should().BeEquivalentTo(expected);
```

and use it like so:

```csharp
DisjointSetShouldBe(parents, [["1", "2", "3", "4", "5", "6"]]); // Merge of 1-height tree and 2-height tree
```

...focusing on the essence of the test.

Reply:
That's nice! I made the change, with two modifications:
- I called it `DisjointSetsShouldBe`, because the sets that are asserted are multiple, not a single one
- I inverted the two parameters, since in assertions the expected value usually comes first

Reply:
Curious: Isn't it usually `actual.ShouldBe.expected`?

Reply:
You are right about the naming, but you still called it `DisjointSetShouldBe`, as in singular. :)

Reply:
> Curious: Isn't it usually `actual.ShouldBe.expected`?

Indeed, but here I can't use a fluent notation unless I create a static class of extension, just to make `DisjointSetsShouldBe` an extension method of `parents`. Which would be incorrect, because it's not the actual that should be something, but a computation of it that should be something.

So we are creating more semantic issues for solving some visual clutter.

I have changed it to `AssertDisjointSets`. That should be consistent and in line with expected order of parameters.